kaiax/staking/staking_info.go

```diff
@@ -0,0 +1,215 @@
+	PoCAddr common.Address `json:"PoCAddr"` // = KIFAddr
+	KCFAddr common.Address `json:"kcfAddr"` // = KEFAddr
+	KFFAddr common.Address `json:"kffAddr"` // = KIFAddr
+}
+
+// StakingInfoResponse is the response type for APIs
+type StakingInfoResponse struct {
+	StakingInfoLegacy
+
+	UseGini bool    `json:"useGini"` // Whether the gini coefficient is used at the requested block number
+	Gini    float64 `json:"gini"`    // The gini coefficient at the requested block number. Returned regardless of `UseGini` value.
+}
+
+func (si *StakingInfo) ConsolidatedNodes() []consolidatedNode {
+	if si.consolidatedNodes == nil {
+		si.consolidatedNodes = si.consolidateNodes()
+	}
+	return *si.consolidatedNodes
+}
+
+func (si *StakingInfo) consolidateNodes() *[]consolidatedNode {
+	// because Go map is not ordered, rList keeps track of the occurrence order of RewardAddrs.
+	// to later arrange the consolidatedNodes.
+	cmap := make(map[common.Address]*consolidatedNode)
+	rList := make([]common.Address, 0, len(si.RewardAddrs))
+
+	for i := range si.NodeIds {
+		r := si.RewardAddrs[i]
+		if cn, ok := cmap[r]; ok {
+			cn.NodeIds = append(cn.NodeIds, si.NodeIds[i])
+			cn.StakingContracts = append(cn.StakingContracts, si.StakingContracts[i])
+			cn.StakingAmount += si.StakingAmounts[i]
+		} else {
+			cmap[r] = &consolidatedNode{
+				NodeIds:          []common.Address{si.NodeIds[i]},
+				StakingContracts: []common.Address{si.StakingContracts[i]},
+				RewardAddr:       r,
+				StakingAmount:    si.StakingAmounts[i],
+			}
+			rList = append(rList, r)
+		}
+	}
+
+	carr := make([]consolidatedNode, 0, len(cmap))
+	for _, r := range rList {
+		carr = append(carr, *cmap[r])
+	}
+	return &carr
+}
+
+// Returns the Gini coefficient among the staking amounts that are greater than or equal to minStake.
+// The amounts are first consolidated by RewardAddr, filtered by minStake, and then summarized to Gini.
+func (si *StakingInfo) Gini(minStake uint64) float64 {
+	// Cache hits only if the same minStake is used.
+	if si.cachedGini == nil || si.cachedGiniMinStake != minStake {
+		g := si.computeGini(minStake)
+		si.cachedGini = &g
+		si.cachedGiniMinStake = minStake
+	}
+	return *si.cachedGini
+}
+
+func (si *StakingInfo) computeGini(minStake uint64) float64 {
+	cnodes := si.ConsolidatedNodes()
+	amounts := make(sort.Float64Slice, 0, len(cnodes))
+
+	for _, cnode := range cnodes {
+		if cnode.StakingAmount >= minStake {
+			amounts = append(amounts, float64(cnode.StakingAmount))
+		}
+	}
+
+	return computeGini(amounts)
+}
+
+func computeGini(amounts sort.Float64Slice) float64 {
+	if len(amounts) == 0 {
+		return EmptyGini
+	}
+
+	sort.Sort(amounts)
+
+	sumOfAbsoluteDifferences := float64(0)
+	subSum := float64(0)
+
+	for i, x := range amounts {
+		temp := x*float64(i) - subSum
+		sumOfAbsoluteDifferences = sumOfAbsoluteDifferences + temp
+		subSum = subSum + x
+	}
+
+	result := sumOfAbsoluteDifferences / subSum / float64(len(amounts))
```

Comment:
I know that you're not the first author of this calculator, but I have a mystery to decrypt how it is the same behavior compared to origin [equation here](https://www.statsdirect.com/help/generatedimages/equations/equation154.svg)

Please take a look the first equation (ignore please the second one).
If I follow the first equation, the code will be changed as:
```diff
diff --git a/kaiax/staking/staking_info.go b/kaiax/staking/staking_info.go
index 9183f4be..de2bf8cf 100644
--- a/kaiax/staking/staking_info.go
+++ b/kaiax/staking/staking_info.go
@@ -167,13 +167,21 @@ func computeGini(amounts sort.Float64Slice) float64 {
        sumOfAbsoluteDifferences := float64(0)
        subSum := float64(0)

+       mean := float64(0)
+       for _, v := range amounts {
+               mean += v
+       }
+       mean /= float64(len(amounts))
+
        for i, x := range amounts {
-               temp := x*float64(i) - subSum
+               // temp := x*float64(i) - subSum
+               temp := x*float64(i+1) - (float64(i+1) * mean)
                sumOfAbsoluteDifferences = sumOfAbsoluteDifferences + temp
                subSum = subSum + x
        }

-       result := sumOfAbsoluteDifferences / subSum / float64(len(amounts))
+       // result := sumOfAbsoluteDifferences / subSum / float64(len(amounts))
+       result := (2 * sumOfAbsoluteDifferences) / mean / float64(len(amounts)) / float64(len(amounts))
        result = math.Round(result*100) / 100
        return result
 }
```

After writing the same behavior based on the equation attached, I noticed that this implementation may be improved(efficient) code getting mean value by not iterating the list, but it still might a mystery how the `subSum` canceled out `2/n`.

Please ignore this question if this takes you too long. I don't want to bother you about this out of scope. Thanks